### ISSUE TITLE
Positive `TypedDict` key membership does not make key access safe

### ISSUE THREAD

**Issue by dsfaccini** [astral-sh/ty#3461]
hey guys, I'm using this fork https://github.com/dsfaccini/ruff/compare/main...codex/ty-pydantic-compat to type check `pydantic-ai` locally. It has patches for all the diagnostics we get from `ty` in `pydantic-ai`

during the process of writing the patches and getting our errors to zero (from 540) I had codex take notes and now I'm going through those notes with subagents to open issues on the more impactful changes (trying to make sure as best as we can that it is actually a valid issue and that there's not already an issue tracking it; if there is I apologize 😅)

sending you best and thank you for your work!
David

Ps. I'm very excited about `ty` so I'll be tracking new diagnostics in our repo https://github.com/pydantic/pydantic-ai/pull/5428 going forward, so we can test the patches long-term

---

A positive literal-key membership check on a `TypedDict` union does not make access to that key safe.

The branch should not necessarily narrow the whole value to the `TypedDict` variant that declares the key, since open `TypedDict`s can have extra keys. But after `"bytes" in file_content`, indexing `file_content["bytes"]` should be safe.

Playground link: https://play.ty.dev/fa50d140-6a86-4415-9069-c0ca40164ffd

### Repro (no external deps)

```python
from typing import TypedDict


class FileWithBytes(TypedDict):
    bytes: bytes


class FileWithUri(TypedDict):
    uri: str


def get_bytes(file_content: FileWithBytes | FileWithUri) -> bytes:
    if "bytes" in file_content:
        return file_content["bytes"]
    raise ValueError
```

Run:

```bash
uvx ty check ty_repro_typed_dict_key_membership.py
```

### Actual result

```text
error[invalid-key]: Unknown key "bytes" for TypedDict `FileWithUri`
  --> ty_repro_typed_dict_key_membership.py:14:16
   |
14 |         return file_content["bytes"]
   |                ------------ ^^^^^^^ Unknown key "bytes"
   |                |
   |                TypedDict `FileWithUri` in union type `FileWithBytes | FileWithUri`
   |
```

### Expected result

No diagnostic. In the positive branch, the membership check establishes that the key exists, so literal key access should be accepted.

### Notes

ty 0.0.35 via `uvx ty check`.

A Pyright-style filter to only the union members that declare the key would be too strong for open `TypedDict`s: an open `FileWithUri` value may legally contain an extra `bytes` key. A more precise behavior would be to narrow the positive branch to something like the original type intersected with a synthesized `TypedDict` requiring key `bytes`, so key access is safe without pretending the value must be `FileWithBytes`.

**Comment by ibraheemdev** [comment]
Thanks for the report. We do support narrowing on `TypedDict` key membership, but it looks like we do not currently support the same for unions of `TypedDict`s. It seems pyright is the only type checker that supports this:

```py
from typing import TypedDict

class FileWithBytes(TypedDict):
    bytes: bytes

class FileWithUri(TypedDict):
    uri: str

def _(file_content: FileWithBytes | FileWithUri):
    if "bytes" in file_content:
        reveal_type(file_content) # pyright: FileWithBytes, mypy/ty/pyrefly: FileWithBytes | FileWithUri
        reveal_type(file_content["bytes"]) # pyright: bytes, mypy/ty/pyrefly: error: [invalid-key]
```

This is also slightly different from https://github.com/astral-sh/ty/issues/1479, which we already support.

**Comment by ibraheemdev** [comment]
I think this is technically only sound for closed `TypedDict`s, but it might be reasonable to ignore that and perform the narrowing unsoundly (as pyright does), unless explicitly annotated with `extra_items`.

**Comment by dsfaccini** [comment]
hey @ibraheemdev thanks for checking this out

Idk if you're commenting for my sake or for your coding agent (I do that a lot)

just wanted to say, let me know if you need anything from my side, I'd be happy to have my bots open a PR if you prefer that

**Comment by AlexWaygood** [comment]
@ibraheemdev I endorse the proposal in the final paragraph of the OP. Rather than filtering out any TypedDicts from the union the way pyright does (which would indeed be unsound for open TypedDicts), we could synthesize a new TypedDict with a single required key and intersect the union with that. I believe that would be fully sound.

**Comment by ibraheemdev** [comment]
@AlexWaygood would the revealed type for `td["bytes"]` be `bytes | Unknown` in that case?

**Comment by AlexWaygood** [comment]
I think it would be `object` (which users might not love, but it would be sound): the subscript access of `"bytes"` on `FileWithBytes & <synthesized TypedDict with key 'bytes'>` would be `bytes & object`, which resolves to `bytes`, but the subscript access of `"bytes"` on `FileWithUri & <synthesized TypedDict with key 'bytes'>` would be `object`, so the subscript access of `"bytes"` on `(FileWithBytes & <synthesized TypedDict with key 'bytes'>) | (FileWithUri & <synthesized TypedDict with key 'bytes'>)` would be `bytes | object`, which just simplifies to `object`

**Comment by charliermarsh** [comment]
Here's a POC: https://github.com/astral-sh/ruff/pull/25188

**Comment by AlexWaygood** [comment]
> Here's a POC: [astral-sh/ruff#25188](https://github.com/astral-sh/ruff/pull/25188)

that definitely gets rid of a lot of ecosystem results, but it doesn't look like the fully sound version that I was proposing 😄

**Comment by charliermarsh** [comment]
Yeah, because it doesn't narrow to `object`. Is narrowing to `object` even useful though?

**Comment by AlexWaygood** [comment]
It's useful in that it enforces type safety and warns you against possible errors, which is the whole point of a type checker!

Part of the reason I think we should go for `object` here is because it's consistent with what we do for `hasattr()` narrowing, for example -- the situation is exactly analogous:

```py
from typing import TypedDict

class Foo:
    x: int

class Bar: ...

def test(obj: Foo | Bar):
    if hasattr(obj, "x"):
        reveal_type(obj.x)  # `object`, because `obj` could be an instance of a `B` subclass
                            # that adds an `x` attribute with an arbitrary type

class Spam(TypedDict):
    x: int

class Ham(TypedDict): ...

def test2(obj: Spam | Ham):
    if "x" in obj:
        reveal_type(obj["x"])  # should also be `object`, I think, because `obj` could be an
                               # instance of a subclass of `Ham` that adds an `x` key with an
                               # arbitrary type
```

**Comment by AlexWaygood** [comment]
and you can narrow the type of `obj["x"]` in further followup `if isinstance(obj["x"], int)` calls later if you need to establish that it is in fact an `int` -- we support subscript narrowing.

**Comment by charliermarsh** [comment]
Makes sense! I updated the branch to use `object`, so you can see some of the newly-added diagnostics from that. It looks pretty good to me. The massive number of removals are from Pydantic, where they have a significant amount of code with this kind of pattern:

```python
    elif schema_type == 'function-before':
        if 'schema' in schema:
            traverse_schema(schema['schema'], context)
        if 'json_schema_input_schema' in schema:
            traverse_schema(schema['json_schema_input_schema'], context)
```

**Comment by AlexWaygood** [comment]
Ah, I _have_ just remembered that intersecting with `TypedDict` types is quite problematic, because they are (though we don't yet recognise it) disjoint from `dict` types, so we'll end up with lots of intersections simplifying to `Never` once we implement that disjointness relation if we intersect with `TypedDict`s too freely. So it's probably better to instead intersect with a synthesized protocol that defines a single `__getitem__` member, looking like this:

```py
class _(Protocol):
    def __getitem__(self, key: Literal["bytes"], /) -> object: ...
```

That should™️ have the same effect.

**Comment by charliermarsh** [comment]
> Is narrowing to object even useful though?

(The reason I asked this is to understand if implementing it was helpful for users, or just something we _could_ do. Your follow-up message was very useful in answering that question.)
